Qemu boot tests

DaanDeMeyer opened the pull request:
Even with the add/remove vmlinuz fix Arch BIOS is still broken. Putting this up already just to see how badly things are broken in the other distros. I'm pretty sure either Mageia or OpenMandriva are going to need fixes as well since I remember one of those not playing well with systemd's gpt-auto-generator.

DaanDeMeyer commented:
We'll definitely have to move all this stuff into pytest soon since it's already a lot of duplication but let's do it like this for now so we can catch some final bugs and get a release out.

DaanDeMeyer commented:
So this actually works as long as you give QEMU enough time (no KVM on Github Actions unfortunately).

Bad thing is lots of stuff is broken. I'll try to fix the easy stuff and I guess we disable the rest for now. Hopefully we'll get contributions that make more of these work in the long run.

DaanDeMeyer commented:
Fixed a ton of bugs but only Arch, Fedora and CentOS work fully on all three bootable setups. I added some extra early failures to quit early when users try to run an unsupported configuration. My idea is to not support anything we can't get running on CI.

DaanDeMeyer commented:
Dropped the removal of quiet since that should be a discussion on its own.

behrmann commented on mkosi:
```diff
@@ -2417,16 +2423,56 @@ def install_arch(args: CommandLineArguments, root: str, do_run_build_script: boo
+                    [Action]
+                    Description = Adding vmlinuz to /boot...
+                    When = PostTransaction
+                    Exec = /bin/bash -c 'while read -r f; do install -Dm644 "$f" /boot/vmlinuz-$(basename $(dirname "$f")); done'
```
```suggestion
                    Exec = /bin/bash -c 'while read -r f; do install -Dm644 "$f" "/boot/vmlinuz-$(basename "$(dirname "$f")")"; done'
```
In the case of `f` containing spaces this would be the full quoting to be safe.

behrmann commented on mkosi:
```diff
@@ -2417,16 +2423,56 @@ def install_arch(args: CommandLineArguments, root: str, do_run_build_script: boo
+                    [Action]
+                    Description = Removing vmlinuz from /boot...
+                    When = PreTransaction
+                    Exec = /bin/bash -c 'while read -r f; do rm -f /boot/vmlinuz-$(basename $(dirname "$f")); done'
```
```suggestion
                    Exec = /bin/bash -c 'while read -r f; do rm -f "/boot/vmlinuz-$(basename "$(dirname "$f")")"; done'
```

behrmann commented on mkosi:
```diff
@@ -2736,17 +2780,30 @@ def install_grub(args: CommandLineArguments, root: str, loopdev: str, grub: str)
         def jj(line: str) -> str:
             if line.startswith("GRUB_CMDLINE_LINUX="):
                 return grub_cmdline
+            elif "GRUB_TERMINAL_INPUT" in line and args.qemu_headless:
+                return "GRUB_TERMINAL_INPUT=\"console serial\""
+            elif "GRUB_TERMINAL_OUTPUT" in line and args.qemu_headless:
```
I'd switch the two `and` clauses, because if the first is false the other will not be evaluated anymore. Not that it makes much of a difference, but it should be quicker to check the truthiness if a bool that to check whether a string is a substring of another.

behrmann commented on mkosi:
```diff
@@ -4644,12 +4703,21 @@ def load_args(args: CommandLineArguments) -> CommandLineArguments:
         args.kernel_command_line.append("console=ttyS0")
 
+    if args.bootable and args.distribution in (Distribution.opensuse, Distribution.mageia):
+        # TODO: Remove once dracut in all these distros is more recent.
```
Maybe leave a comment here which version would be required.

DaanDeMeyer commented on mkosi:
```diff
@@ -1420,6 +1420,11 @@ def prepare_tree(args: CommandLineArguments, root: str, do_run_build_script: boo
 
+        if args.bootable and args.distribution == Distribution.mageia:
+            with open(os.path.join(dracut_dir, "51-mkosi-override-mageia.conf"), "w") as f:
+                f.write("hostonly=no\n")
```
Last one found applies right? Mageia has 50-mageia.conf in /etc/dracut.conf.d that sets hostonly=yes so we override it again to no

behrmann replied:
This is a terrible case of whack-a-mole, but maybe promoting `30-mkosi-hostonly` to `55-mkosi-hostonly` would leave us with a single file doing one thing, but I fear at some point this might not be resolvable for all distros anymore.

DaanDeMeyer replied:
We have to add a separate file anyway for to override omit_dracutmodules again. Might as well to hostonly there as well.

DaanDeMeyer commented on mkosi:
```diff
@@ -2345,6 +2350,7 @@ def install_arch(args: CommandLineArguments, root: str, do_run_build_script: boo
                             kernel_version=()
                             for f in /usr/lib/modules/*/vmlinuz
                             do
+                                [[ -e "$f" ]] || continue
```
See commit message, when a glob doesn't have any matches, it returns the glob itself. That's what we're handling here.

DaanDeMeyer replied:
I'll add a comment and check against the glob itself which should be clearer.

behrmann replied:
The better solution, since we are bash only, would be to set `shopt -s nullglob` (if I remember the shopt synxtax correctly) which makes empty globs truly empty.

Werkov commented on mkosi:
```diff
@@ -4688,6 +4683,10 @@ def load_args(args: CommandLineArguments) -> CommandLineArguments:
         args.kernel_command_line.append("console=ttyS0")
 
+    if args.bootable and args.distribution in (Distribution.opensuse, Distribution.mageia):
+        # TODO: Remove once dracut 045 is available in all these distros.
```
dracut is already >045 in Tumbleweed ([also in Leap 15.2](https://software.opensuse.org/package/dracut), however, I'm only looking after TW mkosi variant). Perhaps this special case is unnecessary then.

DaanDeMeyer commented:
Added opensuse passwordless root login support, adapted CI to use it and removed the dracut workaround for opensuse.

@Werkov Might be a stretch but would be great if we could get the passwordless root login change upstreamed. I can totally understand not wanting to do that though. Also, do you have any idea if it's possible to get opensuse's initramfs/vmlinuz installation scripts to follow the boot loader spec? Even better would be if we could get it to use systemd's kernel-install but that might be a stretch as well.

DaanDeMeyer commented:
Added one more commit to only install dracut config files when necessary.

Werkov commented:
Glad that `nullok` adjustment worked.

Ad upstreaming: patches are welcomed :-) or the topic can be discussed on the [openSUSE mailing list](https://lists.opensuse.org/opensuse-factory/). (My personal stance is that the PAM modification isn't worth it due to its security implications. Ad bootloader specification, I'm not familiar with that and benefits of integration aren't obvious to me.)

Back to this PR, more generally -- this CI suite tests not only mkosi but basically also each distribution's minimalistic image itself. Could you sum up the purpose in the "CI: Add QEMU boot tests" commit?

DaanDeMeyer commented:
I reported the squashfs issue with systemd (https://github.com/systemd/systemd/issues/16818), I need to find some more time to find out exactly what's going wrong there. I disabled that build for now so we can get this merged.

Werkov commented:
@DaanDeMeyer Thanks for the added explanation, it's helpful summary of the expected test behavior.

DaanDeMeyer commented:
Let's merge, hopefully we'll be able to run all distros without exceptions int the future.
